### PR TITLE
add ordering optimization

### PR DESCRIPTION
**Pull request by ttt-77**
This PR implements the ordering of ML execution:
1. Only the inference based on blob will be reordered.
2. The order is determined by **engine_cost** * **selectivity**, where selectivity is calculated as the sum of each blob’s proxy score (0 or 1) divided by the total number of blobs.

### CODE REVIEW

File: tests/inference_service_utils/http_inference_service_setup.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ async def inference(inp: Request):
       service_name = inp.url.path.split('/')[-1]
       inp = await inp.json()
       df = name_to_df[service_name]
-
+      df = df.fillna('Missing')
```
Comment (ttt-77) [inline tests/inference_service_utils/http_inference_service_setup.py:53]: Some values are missing, causing the entire row to be dropped. As a result, the AIDB results differ from the ground truth.

**Comment by ttt-77** [comment]
Could you please review this PR? @ddkang

File: aidb/engine/full_scan_engine.py [hunk 2]
```diff
@@ -1,25 +1,51 @@
...
     Executes a query by doing a full scan and returns the results.
     '''
     # The query is irrelevant since we do a full scan anyway
+  
     is_udf_query = query.is_udf_query
     if is_udf_query:
       query.check_udf_query_validity()
       dataframe_sql, query = query.udf_query
-
+
```
Comment (ddkang) [inline aidb/engine/full_scan_engine.py:28]: whitespace

File: tests/inference_service_utils/http_inference_service_setup.py (hunk 1 above)
Comment (ddkang) [reply tests/inference_service_utils/http_inference_service_setup.py:53]: This doesn't seem like the correct solution

Comment (ddkang) [reply tests/inference_service_utils/http_inference_service_setup.py:53]: What if the value is "missing"?

File: aidb/engine/full_scan_engine.py [hunk 3]
```diff
@@ -1,25 +1,51 @@
...
+    get_currently_supported_filtering_predicates_for_ordering,
+    reorder_inference_engine)
+from sqlalchemy.sql import text
 
 RETRIEVAL_BATCH_SIZE = 10000
 
 
-class FullScanEngine(BaseEngine):
+          
+class FullScanEngine(TastiEngine):
```
[the remark this answers is not in the document]
Comment (ttt-77) [reply aidb/engine/full_scan_engine.py:17]: I think so. Currently, we have a dependency chain: BaseEngine -> TastiEngine -> FullScanEngine, as FullScanEngine uses the proxy score generated by TastiEngine, if it is initialized.

**Review by ddkang — APPROVED** [review]

**Pull request merged**